sdk/python/jobs/pipelines/2b_train_cifar_10_with_pytorch/train-model/main.py

```diff
@@ -123,27 +164,42 @@ def main(args):
     )
 
     # train the model
-    for epoch in range(args.epochs):
-        print("Rank %d: Starting epoch %d" % (rank, epoch))
-        if distributed:
-            train_sampler.set_epoch(epoch)
-        model.train()
-        train(
-            train_loader,
-            model,
-            criterion,
-            optimizer,
-            epoch,
-            device,
-            args.print_freq,
-            rank,
-        )
-
+    valid_min_loss= float('inf')
+    best_model= copy.deepcopy(model)
+    with mlflow.start_run() as run:
+        for epoch in range(args.epochs):
+            print("Rank %d: Starting epoch %d" % (rank, epoch))
+            if distributed:
+                train_sampler.set_epoch(epoch)
+            model.train()
+            # Start the run, log metrics, end the run
+
+            train_loss, valid_loss=train(
+                train_loader,
+                valid_loader,
+                model,
+                criterion,
+                optimizer,
+                epoch,
+                device,
+                args.print_freq,
+                rank,
+            )
+            mlflow.log_metric('train loss',train_loss, step=epoch )
+            mlflow.log_metric('validation loss',valid_loss, step=epoch )
+            
+            if valid_min_loss>=valid_loss:
+                valid_min_loss=valid_loss
+                print("Rank %d: Found min valid loss" % (rank))
+                best_model=copy.deepcopy(model)
```

Comment:
because we're trying to demonstrate best practices in this repo, we may want a different strategy than keeping all models around in memory. hm